central/authprovider/datastore/datastore_impl.go

```diff
@@ -55,23 +55,37 @@ func (b *datastoreImpl) UpdateAuthProvider(ctx context.Context, authProvider *st
 	b.lock.Lock()
 	defer b.lock.Unlock()
 
-	exists, err := b.storage.Exists(ctx, authProvider.GetId())
-	if err != nil {
+	if err := b.verifyExistsAndMutable(ctx, authProvider.GetId(), false); err != nil {
 		return err
 	}
-	if !exists {
-		return errox.NotFound.Newf("auth provider with id %q was not found", authProvider.GetId())
-	}
 	return b.storage.Upsert(ctx, authProvider)
 }
 
 // RemoveAuthProvider removes an auth provider from bolt
-func (b *datastoreImpl) RemoveAuthProvider(ctx context.Context, id string) error {
+func (b *datastoreImpl) RemoveAuthProvider(ctx context.Context, deleteReq *storage.DeleteByIDWithForce) error {
 	if ok, err := authProviderSAC.WriteAllowed(ctx); err != nil {
 		return err
 	} else if !ok {
 		return sac.ErrResourceAccessDenied
 	}
 
+	id := deleteReq.GetId()
+	if err := b.verifyExistsAndMutable(ctx, id, deleteReq.GetForce()); err != nil {
+		return err
+	}
 	return b.storage.Delete(ctx, id)
 }
+
+func (b *datastoreImpl) verifyExistsAndMutable(ctx context.Context, id string, force bool) error {
+	provider, exists, err := b.storage.Get(ctx, id)
+	if err != nil {
+		return err
+	}
+	if !exists {
+		return errox.NotFound.Newf("auth provider with id %q was not found", id)
+	}
+	if provider.GetTraits().GetMutabilityMode() == storage.MutabilityMode_ALLOW_FORCED && !force {
```

Comment:
See my comment below on potential confusion arising from how the checks are performed. Here it's a bit less of an issue because it's directly related to the value of `force`, but for protecting (semi-)immutable objects I would still say that any check should _always_ start with checking if it is allowed or not (`if ... == storage.MutabilityMode_ALLOW { return nil }`), and then any value that is not `ALLOW` but still allows deletion to be performed under certain circumstances is checked in its separate condition.